pkg/compactor/blocks_cleaner.go

```diff
@@ -129,54 +137,175 @@ func NewBlocksCleaner(
+	defer close(deleteChan)
+
+	go func() {
+		c.runActiveUserCleanup(ctx, usersChan)
+	}()
+	go func() {
+		c.runDeleteUserCleanup(ctx, deleteChan)
+	}()
+
+	for {
+		select {
+		case <-t.C:
+			activeUsers, deletedUsers, err := c.scanUsers(ctx)
+			if err != nil {
+				level.Error(c.logger).Log("msg", "failed to scan users blocks cleanup and maintenance", "err", err.Error())
+				c.runsFailed.WithLabelValues(deletedStatus).Inc()
+				c.runsFailed.WithLabelValues(activeStatus).Inc()
+				continue
+			}
+			cleanJobTimestamp := time.Now().Unix()
+			usersChan <- &cleanerJob{
+				users:     activeUsers,
+				timestamp: cleanJobTimestamp,
+			}
+			deleteChan <- &cleanerJob{
+				users:     deletedUsers,
+				timestamp: cleanJobTimestamp,
+			}
+
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (c *BlocksCleaner) runActiveUserCleanup(ctx context.Context, jobChan chan *cleanerJob) {
+	for job := range jobChan {
+		if job.timestamp < time.Now().Add(-c.cfg.CleanupInterval).Unix() {
+			level.Warn(c.logger).Log("Active user cleaner job too old. Ignoring to get recent data")
+			continue
+		}
+		c.cleanUpActiveUsers(ctx, job.users, false) //nolint:errcheck
+	}
 }
 
-func (c *BlocksCleaner) runCleanup(ctx context.Context, firstRun bool) {
-	level.Info(c.logger).Log("msg", "started blocks cleanup and maintenance")
-	c.runsStarted.Inc()
+func (c *BlocksCleaner) cleanUpActiveUsers(ctx context.Context, users []string, firstRun bool) error {
+	level.Info(c.logger).Log("msg", "started blocks cleanup and maintenance for active users")
+	c.runsStarted.WithLabelValues(activeStatus).Inc()
+
+	err := concurrency.ForEachUser(ctx, users, c.cfg.CleanupConcurrency, func(ctx context.Context, userID string) error {
+		userLogger := util_log.WithUserID(userID, c.logger)
+		userBucket := bucket.NewUserBucketClient(userID, c.bucketClient, c.cfgProvider)
+		errChan := make(chan error, 1)
+		defer func() {
+			errChan <- nil
+		}()
+		return errors.Wrapf(c.cleanUser(ctx, userLogger, userBucket, userID, firstRun), "failed to delete blocks for user: %s", userID)
+	})
 
-	if err := c.cleanUsers(ctx, firstRun); err == nil {
-		level.Info(c.logger).Log("msg", "successfully completed blocks cleanup and maintenance")
-		c.runsCompleted.Inc()
-		c.runsLastSuccess.SetToCurrentTime()
+	if err == nil {
+		level.Info(c.logger).Log("msg", "successfully completed blocks cleanup and maintenance for active users")
+		c.runsCompleted.WithLabelValues(activeStatus).Inc()
+		c.runsLastSuccess.WithLabelValues(activeStatus).SetToCurrentTime()
 	} else if errors.Is(err, context.Canceled) {
-		level.Info(c.logger).Log("msg", "canceled blocks cleanup and maintenance", "err", err)
-		return
+		level.Info(c.logger).Log("msg", "canceled blocks cleanup and maintenance for active users", "err", err)
+	} else {
+		level.Error(c.logger).Log("msg", "failed to run blocks cleanup and maintenance for active users", "err", err.Error())
+		c.runsFailed.WithLabelValues(activeStatus).Inc()
```

Comment:
Hmmm for `starting` we are check return and also increase the metric. For the `loop`, we are not doing anything, which then this seems correct. I think we need to make it the same in both places, or increase inline or check failure and increase on caller

Reply:
Addresed